### PR TITLE
refactoring Haddex

### PR DESCRIPTION
**Pull request by metas-rc**
add haddex validations to:
* partner change
* promised date changed
* order completion

https://github.com/metasfresh/metasfresh/issues/10538

### CODE REVIEW

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java [hunk 1]
```diff
@@ -1051,4 +1062,77 @@ public ZoneId getTimeZone(@NonNull final I_C_Order order)
...
+		}
+
+		final org.compiere.model.I_C_BPartner partner = partnerDAO.getById(order.getC_BPartner_ID());
+
+		if (!partner.isHaddexCheck())
+		{
+			return false;
+		}
+
+		if (partner.getDateHaddexCheck() == null)
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:1121]: are we sure this is what we want? what not insisten on a date if isHaddexCheck=true?

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java [hunk 2]
```diff
@@ -1051,4 +1062,77 @@ public ZoneId getTimeZone(@NonNull final I_C_Order order)
...
+
+		validateHaddexDate(order);
+	}
+
+	@Override
+	public void validateHaddexDate(final I_C_Order order)
+	{
+		final IBPartnerDAO partnerDAO = Services.get(IBPartnerDAO.class);
+		final org.compiere.model.I_C_BPartner partner = partnerDAO.getById(order.getC_BPartner_ID());
+		final long differenceBetweenHaddexCheckDateAndPromisedDateInMonths = Math.abs(
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:1094]: why the `abs`? 
not 100% sure what to do if datePromised is before the haddex, check..but should it really depend on how many moths before it is?
maybe we should simply throw the error also if getDatePromised < getDateHaddexCheck ? 
(i guess it's a rare case, so maybe it doesn't make sense to engneer a dedicated error message..)

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:1121]: to be handled in a followup

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:1094]: to be handled in a followup

**Pull request merged**